### ISSUE TITLE
SUBPACKAGE support and dependency removal

### ISSUE THREAD

**Issue by daemonblade** [jrmarino/synth#220]
I'm currently running v3.00, which includes SUBPACKAGE support.

synth builds audio/alsa-plugins as part of upgrading www/chromium. However, after the builds have completed, synth then runs a repository-rebuild and promptly removes the newly built packages:

Stand by, prescanning existing packages.
Removed: alsa-plugins-speex-1.2.7.1_3.pkg
Removed: alsa-plugins-jack-1.2.7.1_3.pkg
Removed: alsa-plugins-ffmpeg-1.2.7.1_3.pkg
Removed: alsa-plugins-samplerate-1.2.7.1_3.pkg
Stand by, recursively scanning 1454 ports serially.
Scanning existing packages.
alsa-plugins-1.2.7.1_3.pkg failed dependency check.
pipewire-0.3.81_2.pkg failed dependency check.
qt6-webengine-6.6.1_5.pkg failed dependency check.
electron28-28.2.2.pkg failed dependency check.
py310-qt6-webengine-6.6.0.pkg failed dependency check.
calibre-7.5.1.pkg failed dependency check.
chromium-121.0.6167.184.pkg failed dependency check.
Packages validated, rebuilding local repository.
Local repository successfully rebuilt

How do I find out why synth is marking alsa-plugin sub-packages as 'out of date'?

**Comment by jrmarino** [comment]
there's probably another place in the code that needs modifying.  I'll take
a look.

**Comment by jrmarino** [comment]
well, here's the cause:
```
audio/alsa-plugins package has less dependencies than the port requires (6)
Query: audio/alsa-lib:alsa-lib-1.2.10_1

alsa-plugins-1.2.7.1_3.pkg failed dependency check.
```

**Comment by daemonblade** [comment]
Cool. I've queried the freebsd-ports@ about this oddity.

**Comment by jrmarino** [comment]
ah - I was thinking about how to update the code to use *_DEPENDS rather than *_DEPENEDS_ALL for this check.  It never occurred to me that it might be a bug on FreeBSD Ports side.

**Comment by jrmarino** [comment]
I think you're right.  These subpackages represent LIB_DEPENDS and it's a bug that the package manifest doesn't list them as dependencies.  I'm going to hold off making changes because I think the framework is the area with the problem.

pkg query -F also confirms that alsa-lib is the only dependency according to the package.

**Comment by daemonblade** [comment]
Unfortunately, the list reports that this is a side effect of the SUBPACKAGE change:
https://lists.freebsd.org/archives/freebsd-ports/2024-February/005414.html

**Comment by jrmarino** [comment]
well, that answer is crap.
"Check dependencies of subpackages instead e.g.," makes no sense.
the issue is the missing dependencies of alsa-plugins.

The makefile lists LIBS dependencies that are not reflected in the final package.
I don't know why the subpackage dependency list is even relevant.

**Comment by jrmarino** [comment]
put another way - If you have an empty system and you install alsa-plugins, you would expect it to pull in the plugins but it won't.  How is that not a bug?

**Comment by jrmarino** [comment]
reading it again, I don't think Jan understood the question.
Maybe I will post to that later.

**Comment by jrmarino** [comment]
at my second attempt I managed to post, but I missed the "Re:" replaced with "Fwd:" so it doesn't show as part of the same thread.
https://lists.freebsd.org/archives/freebsd-ports/2024-February/005426.html

**Comment by jrmarino** [comment]
poudriere has exact same issue with this port:
https://github.com/freebsd/poudriere/issues/1113

**Comment by jrmarino** [comment]
This is kind of a mess.
I can move the extra deps in LIB_DEPENDS_ALL to BUILD_DEPENDS which will fix the master port, but I still need to figure out how to get the expected run dependencies of the subpackages.

basically we have one port creating multiple packages and each package has it's own dependency set.
I have to study the code a little better.
Just guessing, I can build "alsa-plugins-speex" by some command from audio/alsa-plugins.
It's a little tricky.

**Comment by jrmarino** [comment]
actually if it works like ravenports, one build will produce multiple packages.
Per the mail list, I guess the dependency-per-subpackage functionality is not present right now.

**Comment by jrmarino** [comment]
I just thought of another issue.
The package validation is only going to be performed on the "master" port.
Synth doesn't have the concept of multiple packages per directory so it's only going to check the master package, not any of the subpackages.
Maybe this is okay because if the master package exists, the subpackages should exist and be current?

**Comment by jrmarino** [comment]
there's also the possibility the subpackages won't be recognized just deleted as obsolete during the validation checks.

**Comment by jrmarino** [comment]
yes, that's what's probably happening here:
```
Removed: alsa-plugins-speex-1.2.7.1_3.pkg
Removed: alsa-plugins-jack-1.2.7.1_3.pkg
Removed: alsa-plugins-ffmpeg-1.2.7.1_3.pkg
Removed: alsa-plugins-samplerate-1.2.7.1_3.pkg
```
This might be a pretty invasive problem.  If the package manifest identifies these as subpackages then maybe we can just "skip" validating them (assume they are correct).  Otherwise this could be a really big deal.

**Comment by daemonblade** [comment]
The build-dependencies are correct computed with the use of *_DEPENDS_ALL.
For package validation can the code just use *_DEPENDS, as this will be correct?

**Comment by jrmarino** [comment]
No, It's a lot more complicated than that.
I'll get back to trying to figure out how to do this, but it's going to be
one of strategies probably:
1) like flavors, expand the container that holds port information an
additional element per package
2) keep it how it is, but ignore validating subpackages (only validate the
main package)

**Comment by jrmarino** [comment]
I just released v3.01 to address the issues we saw after v3.00.  
It looks like synth was never updated in FreeBSD ports collection, so you'll probably want to update that ticket to v3.01 instead.

I guess subpackages isn't currently being used but synth should support it when it comes back (assuming they don't modify the framework too much when it does).  Let me know if there are problems with subpackages if they occur again.
